### PR TITLE
Expose leaf node with smt.GetLeaf()

### PR DESCRIPTION
**Pull request by roysc**
This will allow access to the contents of a leaf node, including the node path and hashed value.

This will support an [update to Cosmos SDK ADR-040](https://github.com/cosmos/cosmos-sdk/pull/9680) by allowing access to the hashed value and path without any redundant hashing.

Note: should we optimize this to allow fetching the leaf node directly rather than descending the tree, as has been done for value lookup?

### CODE REVIEW

File: smt.go [hunk 1]
```diff
@@ -91,6 +96,54 @@ func (smt *SparseMerkleTree) Get(key []byte) ([]byte, error) {
 	return value, nil
 }
 
+// GetLeaf gets an entire leaf node from the tree.
+func (smt *SparseMerkleTree) GetLeaf(key []byte) (*LeafNode, error) {
+	if bytes.Equal(smt.root, smt.th.placeholder()) {
+		// The tree is empty, return the default value.
+		return nil, nil
```
Comment (adlerjohn) [inline smt.go:103]: I think returning `nil` for the leaf node should be reserved for the error case. If there's no error, it should be a `&LeafNode{Path: path, ValueHash: nil}`. Also, this behavior should be documented both above the function and the struct declaration.

File: smt.go [hunk 2]
```diff
@@ -91,6 +96,54 @@ func (smt *SparseMerkleTree) Get(key []byte) ([]byte, error) {
...
+		leftNode, rightNode := smt.th.parseNode(currentData)
+		if getBitAtFromMSB(path, i) == right {
+			currentHash = rightNode
+		} else {
+			currentHash = leftNode
+		}
+
+		if bytes.Equal(currentHash, smt.th.placeholder()) {
+			// We've hit a placeholder value; this is the end.
+			return nil, nil
```
Comment (adlerjohn) [inline smt.go:132]: Ditto

**Review by adlerjohn — COMMENTED** [review]
It's not clear to me why this is actually needed. Where in https://github.com/cosmos/cosmos-sdk/pull/9680 specifically is it required to get the leaf node?

File: smt.go (hunk 1 above)
Comment (roysc) [reply smt.go:103]: Makes sense :+1:

**Comment by roysc** [comment]
The linked PR is just to change the SMT stored value from the `h(k, v)` to `h(v)` (where `k, v` are the state store KV record). Since the decoupled state store will also store a reverse index mapping the `h(k)` path back to the original `k`, we want a way to get the path without performing a redundant hash on `k`. (Though as I understand it, a change is also planned to map keys directly in the SMT without hashing, in which case that should be unnecessary.)

We also want to use `h(v)` to index values in IPLD (not my domain, @i-norden could shed more light on that), so likewise, we want access to the leaf to avoid hashing `v` again.

**Comment by adlerjohn** [comment]
> we want a way to get the path without performing a redundant hash on `k`
> we want access to the leaf to avoid hashing `v` again.

I 100% guarantee you it's cheaper to perform single hashes than to execute this function (which will perform `log(n)` hashes _and_ disk reads). By several orders of magnitude.

https://github.com/celestiaorg/smt/blob/56045a88b450b04bedd4a8cebabefa73063dce46/smt.go#L103-L148

**Comment by roysc** [comment]
Good point. Even a single disk read might be slower, so optimizing it may also be pointless. I'll confer with Ian and cosmos and see what the actual access pattern will look like.
-------- Original Message --------

**Comment by liamsi** [comment]
> We also want to use h(v) to index values in IPLD

Can you point me to how you use IPLD? I assume you also store the inner nodes in IPFS, too?

**Comment by roysc** [comment]
@liamsi I believe so but I'm not certain - @i-norden would have more info. To my knowledge what we need is to iterate over the SMT's nodes (leaf and probably internal) and retrieve their contents - in which case, this GetLeaf accessor won't be needed anyway. I will close this PR.

**Pull request closed without merging**